add section Upgradeable Contracts (#85)
paritytech/ink-docs: docs/basics/upgradeability.md

            .call_type(
                Call::new()
                    .callee(self.forward_to)
                    .transferred_value(self.env().transferred_value())
                    .gas_limit(0),
            )
            .call_flags(
                ink_env::CallFlags::default()
                    .set_forward_input(true)
                    .set_tail_call(true),
            )
            .fire()
            .unwrap_or_else(|err| {
                panic!(
                    "cross-contract call to {:?} failed due to {:?}",
                    self.forward_to, err
                )
            });
        unreachable!(
            "the forwarded call will never return since `tail_call` was set"
        );
    }
}
```

Using this pattern, you can introduce other message to your proxy contract.
Any messages that are not matched on the proxy contract will be forwarded the specified contract address.

# Updating contract code directly

Following [Substrate's runtime upgradeability](https://docs.substrate.io/tutorials/get-started/forkless-upgrade/) philosophy, ink! also supports an easy way to update your contract code via the special function `set_code_hash()`

> ## Properties

Author: I will contact a designer for this

```rust 
/// Modifies the code which is used to execute calls to this contract address (`AccountId`).
///
/// We use this to upgrade the contract logic. We don't do any authorization here, any caller
/// can execute this method. In a production contract you would do some authorization here.
#[ink(message)]
pub fn set_code(&mut self, code_hash: [u8; 32]) {
    ink_env::set_code_hash(&code_hash).unwrap_or_else(|err| {
        panic!(
            "Failed to `set_code_hash` to {:?} due to {:?}",
            code_hash, err
        )
    });
    ink_env::debug_println!("Switched code hash to {:?}.", code_hash);
}
```

## Storage compatibility

It is developer's responsibility to ensure that the new contract's storage is compatible with the old version.

**You should not change the order in which the contract state variables are declared, nor their type!**

Violating the restriction will not prevent a successful compilation,
but will result in **the mix-up of values** or **failure to read the storage correctly**.
This can be a result of severe errors in the application utilizing the contract.

If the storage of your contract looks like this:
```rust

> use ink_lang as ink;

Reviewer: ```suggestion
```

#[ink(storage)]
pub struct YourContract {
    x: u32,
    y: bool,
}
```

> The procedures listed below will make it ***invalid***:

Reviewer: Not sure how this looks like rendered, but I'd still had a bit of space around
the codeblocks and commentary to make it easier to read in the source

The developer may need to fix a critical bug or introduce a new feature.

For this type of scenario, ink! has different upgrade strategies.
- [Proxy Forwarding](#proxy-forwarding)
  - [Properties](#properties)
- [Replacing Contract Code with `set_code_hash()`](#replacing-contract-code-with-set_code_hash)
  - [Properties](#properties-1)
  - [Storage Compatibility](#storage-compatibility)
  - [A little note on the determinism of contract addresses](#a-little-note-on-the-determinism-of-contract-addresses)
- [Examples](#examples)

## Proxy Forwarding

This method relies on the ability of contracts to proxy calls to other contracts.

### Properties

- Forwards any call that does not match a selector of itself to another contract.
- The other contract needs to be deployed on-chain.
- State is stored in the storage of the contract to which calls are forwarded.

```
User ---- tx ---> Proxy ----------> Implementation_v0
                     |
                      ------------> Implementation_v1
                     |
                      ------------> Implementation_v2
```

Our proxy contract will have these 2 storage fields:

> ```rust

Reviewer 2: ```suggestion

```rust
```

    ///
    /// # Note:
    ///
    /// - We allow payable messages here and would forward any optionally supplied
    ///   value as well.
    /// - If the self receiver were `forward(&mut self)` here, this would not
    ///   have any effect whatsoever on the contract we forward to.
    #[ink(message, payable, selector = _)]
    pub fn forward(&self) -> u32 {
        ink_env::call::build_call::<ink_env::DefaultEnvironment>()
            .call_type(
                Call::new()
                    .callee(self.forward_to)
                    .transferred_value(self.env().transferred_value())
                    .gas_limit(0),
            )
            .call_flags(
                ink_env::CallFlags::default()
                    .set_forward_input(true)
                    .set_tail_call(true),
            )
            .fire()
            .unwrap_or_else(|err| {
                panic!(
                    "cross-contract call to {:?} failed due to {:?}",
                    self.forward_to, err
                )
            });
        unreachable!(
            "the forwarded call will never return since `tail_call` was set"
        );
    }
}
```

:::tip

> Take a look at the selector pattern in the attribute macro: by declaring `selector = _`,

Reviewer 2: ```suggestion
Take a look at the selector pattern in the attribute macro: by declaring `selector = _`
```

we specify that all other messages should be handled by this message selector.

:::

Using this pattern, you can introduce other message to your proxy contract.

> Any messages that are not matched on the proxy contract will be forwarded the specified contract address.

Reviewer 2: ```suggestion
Any messages that are not matched in the proxy contract will be forwarded to the specified contract address.
```

## Replacing Contract Code with `set_code_hash()`

Following [Substrate's runtime upgradeability](https://docs.substrate.io/tutorials/get-started/forkless-upgrade/) 

> philosophy, ink! also supports an easy way to update your contract code via the special function [`set_code_hash()`](https://paritytech.github.io/ink/ink_env/fn.set_code_hash.html)

Reviewer 2: ```suggestion
philosophy, ink! also supports an easy way to update your contract code via the special function [`set_code_hash()`](https://paritytech.github.io/ink/ink_env/fn.set_code_hash.html).
```

### Properties

- Updates the contract code using `set_code_hash()`. This effectively replaces the code which is executed for the contract address.
- The other contract needs to be deployed on-chain.
- State is stored in the storage of the originally instantiated contract.

> Just add the following function to the contract your want to upgrade in future.

Reviewer 2: ```suggestion
Just add the following function to the contract you want to upgrade in the future.
```

```rust 
/// Modifies the code which is used to execute calls to this contract address (`AccountId`).
///
/// We use this to upgrade the contract logic. We don't do any authorization here, any caller
/// can execute this method. In a production contract you would do some authorization here.
#[ink(message)]
pub fn set_code(&mut self, code_hash: [u8; 32]) {
    ink_env::set_code_hash(&code_hash).unwrap_or_else(|err| {
        panic!(
            "Failed to `set_code_hash` to {:?} due to {:?}",
            code_hash, err
        )
    });
    ink_env::debug_println!("Switched code hash to {:?}.", code_hash);
}
```

### Storage Compatibility

> It is developer's responsibility to ensure that the new contract's storage is compatible with the old version.

Reviewer 2: ```suggestion
It is the developer's responsibility to ensure that the new contract's storage is compatible with the storage of the contract that is replaced.
```

:::danger Beware

You should not change the order in which the contract state variables are declared, nor their type!

Violating the restriction will not prevent a successful compilation,
but will result in **the mix-up of values** or **failure to read the storage correctly**.
This can be a result of severe errors in the application utilizing the contract.

:::

> If the storage of your contract looks like this:

Reviewer 2: ```suggestion

If the storage of your contract looks like this:
```

```rust
#[ink(storage)]
pub struct YourContract {
    x: u32,
    y: bool,
}
```

The procedures listed below will make it ***invalid***

Changing the order of variables:

```rust
#[ink(storage)]
pub struct YourContract {
    y: bool,
    x: u32,
}
```

> Removing existing variable:

Reviewer 2: ```suggestion
Removing an existing variable:
```

```rust
#[ink(storage)]
pub struct YourContract {
    x: u32,
}
```

> Changing type of a variable:

Reviewer 2: ```suggestion
Changing the type of a variable:
```
